src/controller/artifact/processor/default.go

```diff
@@ -16,43 +16,117 @@ package processor
 	// try to parse the type from the media type
-	strs := artifactTypeRegExp.FindStringSubmatch(d.mediaType)
+	strs := artifactTypeRegExp.FindStringSubmatch(artifact.MediaType)
 	if len(strs) == 2 {
 		return strings.ToUpper(strs[1])
 	}
 	// can not get the artifact type from the media type, return unknown
 	return ArtifactTypeUnknown
 }
-func (d *defaultProcessor) ListAdditionTypes() []string {
+func (d *defaultProcessor) ListAdditionTypes(ctx context.Context, artifact *artifact.Artifact) []string {
 	return nil
 }
-func (d *defaultProcessor) AbstractMetadata(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error {
-	// do nothing currently
-	// we can extend this function to abstract the metadata in the future if needed
+
+// The default processor will process user-defined artifact.
+// AbstractMetadata will abstract data in a specific way.
+// Annotation keys in artifact annotation will decide which content will be processed in artifact.
+// Here is a manifest example:
+// {
+//   "schemaVersion": 2,
+//   "config": {
+//       "mediaType": "application/vnd.caicloud.model.config.v1alpha1+json",
+//       "digest": "sha256:be948daf0e22f264ea70b713ea0db35050ae659c185706aa2fad74834455fe8c",
+//       "size": 187,
+//       "annotations": {
+//           "io.goharbor.artifact.v1alpha1.skip-list": "metrics,git"
+//       }
+//   },
+//   "layers": [
+//       {
+//           "mediaType": "image/png",
+//           "digest": "sha256:d923b93eadde0af5c639a972710a4d919066aba5d0dfbf4b9385099f70272da0",
+//           "size": 166015,
+//           "annotations": {
+//               "io.goharbor.artifact.v1alphav1.icon": ""
+//           }
+//       },
+//       {
+//           "mediaType": "application/tar+gzip",
+//           "digest": "sha256:d923b93eadde0af5c639a972710a4d919066aba5d0dfbf4b9385099f70272da0",
+//           "size": 166015
+//       }
+//   ]
+// }
+func (d *defaultProcessor) AbstractMetadata(ctx context.Context, artifact *artifact.Artifact, manifest []byte) error {
+	if artifact.ManifestMediaType != v1.MediaTypeImageManifest && artifact.ManifestMediaType != schema2.MediaTypeManifest {
+		return nil
+	}
+	// get manifest
+	mani := &v1.Manifest{}
+	if err := json.Unmarshal(manifest, mani); err != nil {
+		return err
+	}
+
+	// get config layer
+	_, blob, err := d.RegCli.PullBlob(artifact.RepositoryName, mani.Config.Digest.String())
+	if err != nil {
+		return err
+	}
+	defer blob.Close()
+
+	// parse metadata from config layer
+	metadata := map[string]interface{}{}
+	// Some artifact may not have empty config layer
+	if mani.Config.Size != 0 {
+		if err := json.NewDecoder(blob).Decode(&metadata); err != nil {
+			return err
+		}
+	}
+
+	// Populate all metadata into the ExtraAttrs first.
+	artifact.ExtraAttrs = metadata
+
+	annotationController := annotation.NewController()
+	err = annotationController.ProcessAnnotations(artifact, mani, metadata)
```

Comment:
how about do not return error here? It already has the configuration data at hand, in this case UI just doesn't render the icon. 

@steven-zou

Reply:
This error has no relation with icon. Icon will be get in another API.